### PR TITLE
Move around some OOB checks

### PR DESCRIPTION
**Pull request by kasperisager**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -164,37 +164,38 @@ exports.float64 = {
...
       state.end++
     }
   },
   encode (state, b) {
     if (b) {
-      uint.encode(state, b.length)
+      uint.encode(state, b.byteLength)
       state.buffer.set(b, state.start)
-      state.start += b.length
+      state.start += b.byteLength
```
Comment (kasperisager) [inline index.js:177]: I'm actually not entirely sure about this; let me just check how it behaves with non-8-bit typed arrays.

Comment (kasperisager) [reply index.js:177]: It doesn't behave, but that fact was previously masked somewhat. Before, using `b.length`:

```js
cenc.encode(cenc.buffer, Uint32Array.from([1025, 1026, 1027]))
// <Buffer 03 01 02 03>
```

After, using `b.byteLength`:

```js
cenc.encode(cenc.buffer, Uint32Array.from([1025, 1026, 1027]))
// <Buffer 0c 01 02 03 [9 random bytes]>
```

Neither case seems desired, so I'm personally inclined to go with `b.byteLength` and leave the behaviour for anything but `Buffer` and `Uint8Array` undefined. Thoughts?

Comment (mafintosh) [reply index.js:177]: Ah, this actually meant to use .length as it’s an areay of 4 byte entries :)

Comment (mafintosh) [reply index.js:177]: Sorry no confused myself

File: index.js [hunk 2]
```diff
@@ -164,37 +164,38 @@ exports.float64 = {
...
       state.buffer[state.start++] = 0
     }
   },
   decode (state) {
     const len = uint.decode(state)
     if (len === 0) return null
-    const b = state.buffer.subarray(state.start, state.start += len)
-    if (b.length !== len) throw new Error('Out of bounds')
+    if (state.end - state.start < len) throw new Error('Out of bounds')
+    const b = state.buffer.subarray(state.start, state.start + len)
```
Comment (mafintosh) [inline index.js:186]: `return state.buffer.subarray(state.start, state.start += len)` ? should be safe with your check above

File: index.js [hunk 3]
```diff
@@ -252,8 +253,9 @@ exports.string = {
   },
   decode (state) {
     const len = uint.decode(state)
-    const s = b.toString(state.buffer, 'utf8', state.start, state.start += len)
-    if (b.byteLength(s) !== len || state.start > state.end) throw new Error('Out of bounds')
+    if (state.end - state.start < len) throw new Error('Out of bounds')
+    const s = b.toString(state.buffer, 'utf8', state.start, state.start + len)
```
Comment (mafintosh) [inline index.js:257]: return b.toString(..., state.start += len) ?

File: index.js [hunk 4]
```diff
@@ -281,8 +283,9 @@ const fixed = exports.fixed = function fixed (n) {
       state.start += n
     },
     decode (state) {
-      const b = state.buffer.subarray(state.start, state.start += n)
-      if (b.length !== n) throw new Error('Out of bounds')
+      if (state.end - state.start < n) throw new Error('Out of bounds')
+      const b = state.buffer.subarray(state.start, state.start + n)
```
Comment (mafintosh) [inline index.js:287]: same as above

File: index.js (hunk 2 above)
Comment (kasperisager) [reply index.js:186]: Standard doesn't like returning an assignment expression 😕

Comment (mafintosh) [reply index.js:186]: ah i think that's fine if you wrap it in a paren

Comment (kasperisager) [reply index.js:186]: Giving it a shot 👍

Comment (kasperisager) [reply index.js:186]: Still no dice 😔

Comment (mafintosh) [reply index.js:186]: Bummer, we can keep it as is then.

**Pull request merged**